docs/proposals/default-service-plans.md

```diff
@@ -0,0 +1,544 @@
+
+
+``` console
+$ helm install azure/wordpress
+# Provision a new service of type mysql and inject it into Wordpress
+```
+
+
+If the cluster operator has created custom classes, that can be specified when installing the chart:
+
+
+``` console
+$ helm install azure/wordpress \
+    --set className=custom-mysql --set planName=ha-mysql
+```
+
+
+
+## Implementation
+
+A key goal of the implementation is to avoid going off and rewriting or requiring a long term branch or fork, etc. I'd like to see this developed using a feature flag, and making all new fields optional.
+
+
+### Service Class
+
+
+| Field | Description |
+|-------|-------------|
+| Service Type | Examples: mysql, redis|
+| <p>Default Provision Parameters</p><p>Default Bind Parameters</p> | <p>Default parameters to supply to the broker during provisioning and binding.</p> <p>OSB API V3 - Make the schema required, and defaults optional.|
+| Default Bind Response Transform | <p>Default json key mapping to transform a non-standard broker response into something that can be relied upon.</p> <p>For example, if a broker returns "database.name" (a nested value) in the response, it can be changed to "database" to un-nest it and apply a standard name.</p>|
+| Controller Reference |<p>Set by Service Catalog when populating the catalog from a broker list to indicate that it is managed by the controller.</p> <p>When not present, the class is ignored during a relist. This gives operators a way to define custom classes.</p>|
+
+Notes
+
+* Relist needs to take into account if the class is managed by the controller or was created by the cluster operator.
+* Defaults can be defined at the Cluster or Namespace level.
+* For simplification, only broker bind response bodies can be transformed, not provision, since that's where we require that functionality today.
+
+
+### Service Plan
+
+| Field | Description |
+|-------|-------------|
+| Service Type | Examples: mysql, redis |
+| annotation: is-default-plan=true | <p>Specifies whether or not a Service Plan is the default for its Service Type.</p> <p>When annotated as the default and a Service Type is set, this plan is used to dynamically provision an instance for that type.</p> |
+| Controller Reference|<p>Set by Service Catalog when populating the catalog from a broker list to indicate that it is managed by the controller.</p> <p>When not present, the plan is ignored during a relist. This gives operators a way to define custom plans.</p> |
+
+
+Notes:
+
+* The controller will enforce that only one default is set per service type.
+* For simplification, defaults are set at the class level only, plans cannot override them.
+
+
+### Service Instance
+
+
+| Field | Description |
+|-------|-------------|
+| Service Type | Examples: mysql, redis |
+| Provision Parameters | Overrides the default provision parameters defined on the class/plan. |
```

Comment:
These are in addition to parameters? does sc have to look up and recalculate the defaults + provision parameters for each update? Does it have to setup a watcher to watch the defaults so that they can flow to the existing instances? Or does it have a copy of the defaults used.

Reply:
Sorry for the confusion, that was poor naming on my part. I'll fix the name of the field and the [example yaml](https://github.com/carolynvs/service-catalog/blob/default-service-plan-proposal/docs/proposals/default-service-plans.md#provisioning-with-defaults). This is the existing parameters field on the service instance spec.

When a service instance is provisioned, the controller:

1. Selects the appropriate class/plan using either the values specified by the instance, or by the default service type if the instance doesn't specify a plan.
1. The defaults defined on the class and plan are merged with the parameters defined on the instance. The precedence order is Class < Plan < Instance.
1. The final set of defaults are persisted on the instance.

There is no reference or tie back to the objects that provided the defaults. They are only used to **default** the parameters, and then going forward the parameters must be managed on the instance directly, just as they are today.